converter/converter.go

```diff
@@ -0,0 +1,437 @@
+	switch fieldType {
+	case "bool":
+		return "bool"
+	case "string":
+		return "string"
+	case "bytes":
+		return "[]byte"
+	case "double":
+		return "float64"
+	case "float":
+		return "float32"
+	case "int32":
+		return "int"
+	case "sint32", "sfixed32":
+		return "int32"
+	case "int64", "sint64", "sfixed64":
+		return "int64"
+	case "uint32", "fixed32":
+		return "uint32"
+	case "uint64", "fixed64":
+		return "uint64"
+	default:
+		// This is either an unrecognised type, or a custom type.
+		return fieldType
+	}
+}
+
+func (b *builder) handleProtoType(typ proto.Visitee) error {
+	switch typ.(type) {
+	case *proto.Syntax:
+		// Do nothing with syntax
+	case *proto.Package:
+		// This gets translated at the very end because it is used
+		// in conjuction with the option "go_package" when writting
+		// a Gunk package decleration.
+		b.pkg = typ.(*proto.Package)
+	case *proto.Import:
+		// All imports need to be grouped and written out together. This
+		// happens at the end.
+		b.imports = append(b.imports, typ.(*proto.Import))
+	case *proto.Message:
+		b.handleMessage(typ.(*proto.Message))
+	case *proto.Enum:
+		b.handleEnum(typ.(*proto.Enum))
+	case *proto.Service:
+		b.handleService(typ.(*proto.Service))
+	case *proto.Option:
+		o := typ.(*proto.Option)
+		lit := o.Constant
+		value, err := b.handleLiteralString(lit)
+		if err != nil {
+			b.emitError(lit.Position, "expected string literal in top level option: %v", err)
+			return nil
+		}
+		b.pkgOpts = append(b.pkgOpts, pkgOpt{
+			typeName: o.Name,
+			value:    value,
+			comment:  o.Comment,
+		})
+	default:
+		return fmt.Errorf("unhandled proto type %T", typ)
+	}
+	return nil
+}
+
+// handleMessageField will convert a messages field to gunk.
+func (b *builder) handleMessageField(w *strings.Builder, field proto.Visitee) error {
+	var name string
+	var typ string
+	var sequence int
+	var repeated bool
+	var comment *proto.Comment
+
+	switch field.(type) {
+	case *proto.NormalField:
+		ft := field.(*proto.NormalField)
+		name = ft.Name
+		typ = b.goType(ft.Type)
+		sequence = ft.Sequence
+		comment = ft.Comment
+		repeated = ft.Repeated
+	case *proto.MapField:
+		ft := field.(*proto.MapField)
+		name = ft.Field.Name
+		sequence = ft.Field.Sequence
+		comment = ft.Comment
+		keyType := b.goType(ft.KeyType)
+		fieldType := b.goType(ft.Field.Type)
+		typ = fmt.Sprintf("map[%s]%s", keyType, fieldType)
+	default:
+		return fmt.Errorf("unhandled message field type %T", field)
+	}
+
+	if repeated {
+		typ = "[]" + typ
+	}
+
+	// TODO(vishen): Is this correct to explicitly camelcase the variable name and
+	// snakecase the json name???
+	// If we do, gunk should probably have an option to set the variable name
+	// in the proto to something else? That way we can use best practises for
+	// each language???
+	b.Write(w, 1, comment, "%s %s `pb:\"%d\" json:\"%s\"`\n", snaker.ForceCamelIdentifier(name), typ, sequence, snaker.CamelToSnake(name))
```

Comment:
you've made the write method fairly low level, so I'd split these very long lines into multiple, like:

```
b.Write(w, 1, comment, "%s %s", snaker.ForceCamelIdentifier(name), typ)
b.Write(w, 1, comment, " `pb:\"%d\" json:\"%s\"`\n",  sequence, snaker.CamelToSnake(name))
```